import "ydb/public/api/protos/ydb_issue_message.proto";
import "ydb/public/api/protos/ydb_operation.proto";
import "ydb/public/api/protos/ydb_status_codes.proto";
import "ydb/public/api/protos/ydb_table.proto";
import "ydb/public/api/protos/ydb_value.proto";
import "google/protobuf/any.proto";


package NKikimrIndexBuilder;
option java_package = "ru.yandex.kikimr.proto";

message TColumnBuildSetting {
    optional string ColumnName = 1;
    optional Ydb.TypedValue default_from_literal = 2;
    optional bool NotNull = 3;
    optional string Family = 4;
}

message TColumnBuildSettings {
    repeated TColumnBuildSetting column = 1;
    optional string Table = 2;
}

message TIndexBuildSettings {
    optional string source_path = 1;
    optional Ydb.Table.TableIndex index = 2;
    optional TColumnBuildSettings column_build_operation = 7;
    optional bool pg_mode = 8 [ default = false];
    optional bool if_not_exist = 9 [ default = false];

    optional uint32 max_batch_rows = 3 [ default = 500 ];
    optional uint64 max_batch_bytes = 4 [ default = 8388608 ];
    optional uint32 max_shards_in_flight = 5 [ default = 32 ];
    optional uint32 max_retries_upload_batch = 6 [ default = 50 ];

    optional google.protobuf.Any AlterMainTablePayload = 10;
}

message TIndexBuild {
    optional uint64 Id = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
    optional Ydb.Table.IndexBuildState.State State = 3;
    optional TIndexBuildSettings Settings = 4;
    optional float Progress = 5 [default = 0];
}

message TEvCreateRequest {
    optional Ydb.Operations.OperationParams OperationParams = 1;
    optional uint64 TxId = 2;
    optional string DatabaseName = 3;
    optional TIndexBuildSettings Settings = 4;
}

message TEvCreateResponse {
    optional uint64 TxId = 1;
    optional Ydb.StatusIds.StatusCode Status = 2;
    repeated Ydb.Issue.IssueMessage Issues = 3;
    optional TIndexBuild IndexBuild = 4;
    optional sint32 SchemeStatus = 5 [default = -1]; //flat_tx_scheme.proto - enum EStatus, -1 is inexistent enum value
}

message TEvGetRequest {
    optional string DatabaseName = 1;
    optional uint64 IndexBuildId = 2;
}

message TEvGetResponse {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
    optional TIndexBuild IndexBuild = 3;
}

message TEvCancelRequest {
    optional uint64 TxId = 1;
    optional string DatabaseName = 2;
    optional uint64 IndexBuildId = 3;
}

message TEvCancelResponse {
    optional uint64 TxId = 1;
    optional Ydb.StatusIds.StatusCode Status = 2;
    repeated Ydb.Issue.IssueMessage Issues = 3;
}

message TEvForgetRequest {
    optional uint64 TxId = 1;
    optional string DatabaseName = 2;
    optional uint64 IndexBuildId = 3;
}

message TEvForgetResponse {
    optional uint64 TxId = 1;
    optional Ydb.StatusIds.StatusCode Status = 2;
    repeated Ydb.Issue.IssueMessage Issues = 3;
}

message TEvListRequest {
    optional string DatabaseName = 1;
    optional uint64 PageSize = 2;
    optional string PageToken = 3;
}

message TEvListResponse {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
    repeated TIndexBuild Entries = 3;
    optional string NextPageToken = 4;
}

enum EBuildStatus {
    INVALID = 0;
    ACCEPTED = 1; // DEPRECATED don't use it, it does nothing, we cannot remove it for now, because old nodes can send it
    IN_PROGRESS = 2;
    DONE = 3;

    ABORTED = 4;
    BUILD_ERROR = 5;
    BAD_REQUEST = 6;
}

message TEvUploadSampleKResponse {
    optional uint64 Id = 1;

    optional Ydb.StatusIds.StatusCode UploadStatus = 2;
    repeated Ydb.Issue.IssueMessage Issues = 3;

    optional uint64 RowsDelta = 4;
    optional uint64 BytesDelta = 5;
}

